### DISCUSSION TITLE
Implementation of conditionals in dnPhyloCTMC

### DISCUSSION THREAD

**Discussion by ms609** [revbayes/revbayes#619] [Q&A]
I'm looking to implement a conditional in dnPhyloCTMC.  I've had success using `dnInverse()` (#582) for some simple(ish) cases, but am now working through a more complex case where a specific C++-level implementation seems warranted.

To get a handle on what would be involved in an implementation, I've been working through the code for the existing "variable" and "informative" conditions.  In most cases, the code contains separate clauses for each ascertainment bias, but when it comes to computing the corrections (in `PhyloCTMCSiteHomogeneousConditional.h`), this is performed with a single function (`computeTipCorrection` and equivalent for nodes and roots) that does not differentiate between the two conditions.  I can see (I think) how this correction would handle the variable case, but not the informative.  The two are clearly handled differently (see #552), but I've not been able to see - in the code or the debugger - where the informative-specific code lies or how it's triggered.  Am I looking in the wrong place for the implementation of the informative correction, or am I misunderstanding something more fundamental?

**Comment by davidcerny** [comment]
It looks like the two corrections only have different effects in `PhyloCTMCSiteHomogeneousConditional::sumRootLikelihood()` for standard characters and `PhyloCTMCSiteHomogeneousDollo::computeIntegratedNodeCorrection()` for Dollo characters, but the nested loops and compound iterators don't make it easy (for me at least) to understand what's going on – even though there are comments in the source code trying to help. Based on these lines:

```cpp
// number of correction patterns per character state
if (coding & (AscertainmentBias::INFORMATIVE ^ AscertainmentBias::VARIABLE))
{
    numCorrectionPatterns = std::pow(2.0f, float(this->num_chars - 1));
}
```

my impression is that we don't actually simulate more correction patterns under `coding=informative` than under `coding=variable`, contrary to @bredelings's original hypothesis (see #552). It's just that in the former case, we're including some special handling for each correction pattern, which I guess must be what accounts for the difference in runtime and memory consumption between the two treatments.
